### PR TITLE
Remove REVERTcast 

### PR DESCRIPTION
**Pull request by SkySkimmer**
See https://github.com/coq/coq/issues/13985

Overlays:
 - https://github.com/LPCIC/coq-elpi/pull/288
 - https://github.com/QuickChick/QuickChick/pull/234
 - https://github.com/damien-pous/relation-algebra/pull/22
 - https://gitlab.inria.fr/gappa/coq/-/merge_requests/3
 - https://github.com/MetaCoq/metacoq/pull/589
 - https://github.com/mattam82/Coq-Equations/pull/429
 - https://github.com/coq-community/aac-tactics/pull/88

### CODE REVIEW

**Comment by maximedenes** [comment]
I don't quite remember, but the debug printer may also not print these casts. How about adding them there too?

**Review by Zimmi48 — APPROVED** [review]

**Comment by JasonGross** [comment]
How does this factor with the `_ = _ :> _` notation?  (Is this tested in the test-suite?)

**Comment by SkySkimmer** [comment]
>How does this factor with the _ = _ :> _ notation? (Is this tested in the test-suite?)

It seems to work fine. There were already some incidental tests (eg https://github.com/coq/coq/blob/eeb4bc6d0ba5d4e6c0ee77429a055d8e2d957404/test-suite/ssr/view_case.v#L23), I added a specific test.

**Comment by SkySkimmer** [comment]
Quickchick failure unrelated (it can be seen on master: https://gitlab.com/coq/coq/-/jobs/1588561248

**Comment by SkySkimmer** [comment]
ping @coq/pretyper-maintainers @coq/extensible-syntax-maintainers

**Review by herbelin — APPROVED** [review]
The merge of `cast_type` and `cast_kind` is appreciable!

I suggest to add that `Printing All` includes `Printing Tactic Casts` in the doc.

Actually, is it worth to have a special flag only for `REVERTcast`? Moreover, tactics can also add `DEFAULTcast`, so it is a bit strange to have `Tactic` in the name. If a new option is added, why not e.g. to target all casts at once?

Incidentally, some parts require a review from @jfehrle.

**Comment by SkySkimmer** [comment]
It seems always printing casts without an option works OK, or at least does not disturb many output tests. Should we go that way?

**Comment by SkySkimmer** [comment]
Also, what's a good description of the difference between `:` and `:>`?

File: doc/changelog/02-specification-language/14773-revertcast.rst
Comment (jfehrle) [inline doc/changelog/02-specification-language/14773-revertcast.rst:3]:
```suggestion
  Some cast nodes inserted by tactics were previously printed as normal casts (`` c > t``) when
  ``Set Printing All`` was on and otherwise were invisible.
  They are now printed (and may be entered) as ``c :> t`` when ``Set Printing Tactic Casts`` is on
```

It was truly "some"/"certain" and not "all"?  If you can enter `:>` when `Set Printing Tactic Casts` is not set, then for clarity, make "and may be entered" a separate sentence.

File: doc/sphinx/language/core/definitions.rst
Comment (jfehrle) [inline doc/sphinx/language/core/definitions.rst:57]:
```suggestion
   When set, ``:>`` casts are printed.  Otherwise the term is printed with the cast removed.
   The default is unset.
```

File: doc/sphinx/language/core/definitions.rst
Comment (jfehrle) [inline doc/sphinx/language/core/definitions.rst:45]: We try to present things in the order of importance to users while also grouping similar items, to the extent possible.  Should this be #2 in the list?  If not, please reorder in `orderedGrammar` and change the order of the descriptions below to match the new order.

**Review by jfehrle — COMMENTED** [review]
Wording suggestions

**Comment by herbelin** [comment]
> It seems always printing casts without an option works OK, or at least does not disturb many output tests. Should we go that way?

Probably, it happens mostly in proof terms and it is not so common to print proofs.

> Also, what's a good description of the difference between `:` and `:>`?

I checked and it seems that the intended behavior is broken.

In principle, when checking that the type `T` of `t` is a subtype of `U` in `t :> U`, the conversion prioritary tries to reduce `T` before `U`  when there are two head constants and one must be reduced while with `t : U`, it prioritary tries to reduce `U` before `T`.

For instance:
```
Fixpoint ack n := match n with 0 => S | S n => fix ackSn m := match m with 0 => ack n 1 | S m => ack n (ackSn m) end end.
Check fun x : ack 4 4 = 0 => x : id (ack 4 4) = 0. (* should unfold id and be immediate *)
Check fun x : ack 4 4 = 0 => x :> id (ack 4 4) = 0. (* should unfold and compute ack (but is immediate) *)
Check fun x : id (ack 4 4) = 0 => x : ack 4 4 = 0. (* should unfold and compute ack on the right *)
Check fun x : id (ack 4 4) = 0 => x :> ack 4 4 = 0. (* should unfold id and be immediate (but compute ack) *)
```

**Comment by SkySkimmer** [comment]
l2r has been ignored since approximately db80daaf82a08a1475c65f7c82bffb63c7efd27a
revertcast is treated differently in printing, and in typeops.judge_of_cast. AFAICT the later is always called with defaultcast though.

**Comment by SkySkimmer** [comment]
Related https://github.com/coq/coq/pull/14925

**Comment by SkySkimmer** [comment]
> > It seems always printing casts without an option works OK, or at least does not disturb many output tests. Should we go that way?
> 
> Probably, it happens mostly in proof terms and it is not so common to print proofs.
> 
> > Also, what's a good description of the difference between `:` and `:>`?
> 
> I checked and it seems that the intended behavior is broken.
> 
> In principle, when checking that the type `T` of `t` is a subtype of `U` in `t :> U`, the conversion prioritary tries to reduce `T` before `U` when there are two head constants and one must be reduced while with `t : U`, it prioritary tries to reduce `U` before `T`.
> 
> For instance:
> 
> ```
> Fixpoint ack n := match n with 0 => S | S n => fix ackSn m := match m with 0 => ack n 1 | S m => ack n (ackSn m) end end.
> Check fun x : ack 4 4 = 0 => x : id (ack 4 4) = 0. (* should unfold id and be immediate *)
> Check fun x : ack 4 4 = 0 => x :> id (ack 4 4) = 0. (* should unfold and compute ack (but is immediate) *)
> Check fun x : id (ack 4 4) = 0 => x : ack 4 4 = 0. (* should unfold and compute ack on the right *)
> Check fun x : id (ack 4 4) = 0 => x :> ack 4 4 = 0. (* should unfold id and be immediate (but compute ack) *)
> ```

Does it then make sense to use revertcast for reduction tactics?
I mean 
~~~coq 
Lemma foo (x:ack 4 4 = 0) : id (ack 4 4) = 0.
Proof.
  unfold id.
  exact x.
Defined.
~~~
produces the second one, which is only fast because we ignore l2r.

**Comment by SkySkimmer** [comment]
Here's an agressive removal of revertcast infrastructure. The last commit reverts https://github.com/coq/coq/pull/13531 so we may not be able to keep it.

**Comment by SkySkimmer** [comment]
@coqbot why no pipeline?

**Comment by JasonGross** [comment]
Does this mean that `cbv` no longer inserts casts?  Or that the casts it inserts print by default?

**Comment by SkySkimmer** [comment]
It always prints.

**Comment by SkySkimmer** [comment]
Can I get an assignee / merge?

**Comment by Alizter** [comment]
@SkySkimmer Do I need to check/do anything with the overlays?

**Comment by SkySkimmer** [comment]
Tell upstream to merge when you merge

**Comment by Alizter** [comment]
Also some of the overlays are still drafts is this intended?

**Pull request merged**
